System time provider

eranbh opened the pull request:
system time collector pr.
please see README for details.

Commit:
initial commit

1. added basic time collection thread
2. still need to test the cmd
3. still need to profile cmd

Commit:
flow now works end to end (sending timestamp to cli)
started profiling the following:
1. no sync at all (racing on the data containing the time)
2. sync using a pthread metex
still needs to be done:
1. profile with atomic operations (make the increment and load atomic)
2. handle errors

Commit:
fix race in collector thread

Commit:
1. added comments
2. fixed sync issue on timestamp [2 build modes]
3. added a build mode that offers formating
still need to profile

Commit:
fixed timeout issue in multiplexer

Commit:
readme

Commit:
comments on making and readme

Commit:
ifdef out mutex code

Commit:
last minute fixups

grishaf commented on redis-6.0.9/src/modules/systemtime.c:
```diff
@@ -0,0 +1,249 @@
+	///// send the result back.
+#ifdef __USE_HUMAN_TIME	
+	time_t time = system_time_usec / NANO_IN_SEC;
+	RedisModule_ReplyWithCString(ctx, asctime(localtime(&time)));
```
`asctime`  very dangerous to be used, as if few threads using this function they can override eachother.
Form glibc implementation:
```
static char result[                 3+1+ 3+1+20+1+20+1+20+1+20+1+20+1 + 1];

char *
--
asctime (const struct tm *tp)
{
return asctime_internal (tp, result, sizeof (result));
}
```

eranbh replied:
amazing. you were right about this being an issue.
instead of using the library functions, i used code that "parses" the timestamp just for the fields required.
it improved the performs by a factor of almost 4!!!
cmdstat_drivenets.get.time:calls=11,usec=116,usec_per_call=10.55

grishaf commented on redis-6.0.9/src/modules/systemtime.c:
```diff
@@ -0,0 +1,249 @@
+			// use the number of events from the result to account for them
+			uint64_t times_expired;
+			ssize_t size = read(events[0].data.fd, &times_expired, sizeof(uint64_t));
+			if(size != sizeof(uint64_t)) {
```
there are errors that might happen here that should not be logged like `EINTR`.
also should be logged and not printf

eranbh replied:
done:
i added code to avoid logging in these cases.

Commit:
fix possible race on mutex

grishaf commented on redis-6.0.9/src/modules/systemtime.c:
```diff
@@ -0,0 +1,249 @@
+#endif
+		}
+		else{
+			printf("no events happened. stopping multiplexer\n");
```
also here

eranbh replied:
done

grishaf commented on redis-6.0.9/src/modules/systemtime.c:
```diff
@@ -0,0 +1,249 @@
+    // on the read when we leave the epoll on a _non_ timer related event
+	s_timer_fd = timerfd_create(CLOCK_MONOTONIC, TFD_CLOEXEC | TFD_NONBLOCK);
+	if (0 > s_timer_fd){
+		printf ("timerfd_create failed. errno: %d\n", errno);
```
you should abort and not continue

eranbh replied:
done.

grishaf commented on redis-6.0.9/src/modules/systemtime.c:
```diff
@@ -0,0 +1,249 @@
+int start_sys_time_collector_thread()
+{
+    pthread_t threadId = -1;
+    int sts = pthread_create(&threadId, NULL, &collector_thread_work, 0);
```
Great! this was the main Idea

grishaf commented on redis-6.0.9/src/modules/systemtime.c:
```diff
@@ -0,0 +1,249 @@
+	while(1)
+	{
+		// if the first timer didnt start in 40 seconds. something is wrong.
+		int fireEvents = epoll_wait(s_epoll_fd, events, 1, MILLI_IN_SEC * 40);
```
why epoll + timer, and not just sleep with reading the current time and appending the diff from last update?

grishaf commented on redis-6.0.9/src/modules/systemtime.c:
```diff
@@ -0,0 +1,249 @@
+			do{
+				new_sys_time = system_time_usec = s_system_time_usec;
+				new_sys_time += times_expired * (jiffi_in_seconds * NANO_IN_SEC);
+			}while (0 == __sync_bool_compare_and_swap(&s_system_time_usec,
```
why compare and swap? you are the only writer to this value, so atomic inc is not enough?

eranbh replied:
you are super right!
i was looking at the man for: __sync_add_and_fetch(), which is just a simple store, and saw issue with this.
for example: https://stackoverflow.com/questions/6786284/gcc-atomic-built-in-functions

this says that there are actually _2_ operations (both atomic on their own, but _not_ with respect to each other)
someone suggested that if u need a load-old AND load-old+delta, u can just use cas (which as u pointed out, will always succeeds in this case).
wdut?

eranbh replied:
i changed it to a simple __sync_add_and_fetch with the delta.

grishaf commented on redis-6.0.9/src/modules/systemtime.c:
```diff
@@ -0,0 +1,249 @@
+	/* create timer */
+    // CLOCK_MONOTONIC is a good clock to use here, as its
+    // man states that it is _not_ sensetive to changes
+    // we want a non-blocking fd, so that we can avoid blocking
```
it does not sensitive to backward time changes (so it always goes up, but still if time changes it might trigger faster), `CLOCK_MONOTONIC_RAW` is the preferred one, but I think timer is not support this in older kernel version.

eranbh replied:
understood.

grishaf reviewed (changes requested):
Hello Eran, 
thanks for your time.
overall you are in a good direction, however the requested output is in the following format:
`"10-10-2020 12:14"`. This detail is important (both as a feature request, and implementation detail that you can use).

your 40-36us per request is very high result, again check the output that requested .

eranbh commented:
thanks for the feedback!! @grishaf 
ill probably do the fixing later today, and ping u back.

Commit:
parse epoch manually for format string

Commit:
code review comments

1. added redis based logging.
2. changed the sync facility in atomics case from cas to fetch_add
3. added more fine grained error handling to the multiplexer

orenhecht commented:
@eranbh are you done with the fixes? Or do you still need more time? (Reminder you have until end of week)

Commit:
adding profile results for manual broken down time

eranbh commented:
@orenhecht 
Hey,
As per the comments marked "done" - I am done.                                  
There was a suggestion to convert this from a reactive form (using timers) to a proactive form (explicitly calling sleep from my code). I can do that (maybe on a new branch?).
I am curious to know if the performance metrics achieved here are sufficient? (they are very close to hset, on average).

note: the code that does the timestamp parsing is a simplified version of the full algorithm that is needed for this. it is a very complex algorithm (accounting for leap years and skews etc ... ) and i didn't think it was the point of this exercise, so i only took some of it (its not handling all edge cases etc).
thanks!

Commit:
no seconds needed in result

orenhecht commented:
Hi,

The assignment asks for it to be as efficient as possible, if you believe this is the most efficient way and consider it done then fine.

I would like to get your approval for finishing the assignment so we could evaluate it (you still have time if needed).

Oren.

eranbh commented:
I actually thought of another optimization (not sure how beneficial this would be):
I could cache the broken-down time components, and just add to it (whenever there is a "spill over" of one component i adjust the next component).
id like to try it out later today. ill let you know.
thanks.

Commit:
performance gain

by pre-formating the timestamp there was a 2x gain
this only works with a mutex, as i need a way to load/ store
a buffer big enough for the entire pre-formated string, and
the native procedures are not compatible.

Commit:
updating profiling results

Commit:
cleanup

1. updated readme
2. updated profile_results file
3. removed _all_ profiling modes from code [all ifdefs cleaned]

eranbh commented:
Hi All,

I have cleaned up all profiling related code, and ifdefs.
from my profiling, the current situation gives this result (on average):
cmdstat_drivenets.get.time:calls=50,usec=249,usec_per_call=4.98
this comes _very_ close to the measurement i got for sending a raw timestamp back (just a long).
*any* feedback will be greatly appreciated!!!
have a good day.

fyi - 
all the build modes that allow profiling are now in this branch: https://github.com/eranbh/drivenets/tree/system_time_provider_profiling

Commit:
adjust year

orenhecht commented:
Hi @eranbh,

Do you consider this the final submission? Can we evaluate it?

Oren.

eranbh commented:
Hi, @orenhecht 

Yes please.

Thanks!

grishaf commented on redis-6.0.9/src/modules/systemtime.c:
```diff
@@ -0,0 +1,246 @@
+				// the man says not reading a full uint64_t is an error
+				// report an error and terminate, unless its these cases
+				if(errno != EAGAIN && errno != EINTR){
+					RedisModule_Log(ctx, "error", "read error. errno: %d", errno);
```
using redis ctx from another thread:
```
RedisModule_GetDetachedThreadSafeContext
RedisModuleCtx *RedisModule_GetDetachedThreadSafeContext(RedisModuleCtx *ctx);
Return a detached thread safe context that is not associated with any specific blocked client, but is associated with the module's context.

This is useful for modules that wish to hold a global context over a long term, for purposes such as logging.
```

eranbh replied:
understood.

grishaf commented on redis-6.0.9/src/modules/systemtime.c:
```diff
@@ -0,0 +1,246 @@
+ 
+	its.it_interval.tv_sec = 0;
+	its.it_interval.tv_nsec = jiffi_in_seconds * NANO_IN_SEC;
+	its.it_value.tv_sec = 1; // we start aligned to a full second
```
the request was to show time with minute accuracy, so you doing a lot of not needed work here

eranbh replied:
yeah ... that was a bit of an overkill :)

grishaf commented on redis-6.0.9/src/modules/systemtime.c:
```diff
@@ -0,0 +1,246 @@
+	// dont hold the lock for the ipc
+	char buffer[64] = {0};
+	pthread_mutex_lock(&lock);
+	memcpy(buffer, s_formated_time_buffer, 64);
```
why the memcpy to buffer? why not
```
RedisModule_ReplyWithCString(ctx, s_formated_time_buffer);
```

eranbh replied:
i didnt want to hold the mutex while doing the ipc.
i dont have a profile figure for this, so i dont know if this is meaningful in terms of performance (??)

grishaf reviewed (commented):
Hello Eran, thanks a lot for your time, we really appreciate it.
Your last result is very close to what we expected, the idea was to realise that the requested output is in minutes so you can do all the job in background thread and optimise the gets.

One thing I would improve in your solution is to use double buffer array that hold the otuput string to avoid the locks.

Good Job! I've added few comments.
We would consult about the answer and our HR will contact you.
Thanks again

eranbh commented:
@grishaf 
wow! that's an amazing idea!! let me see if i get that right ... 
so there would be **two** buffers and an **int** that chooses between them?
and the collector thread would "publish" the new buffer after every update?
